File: 2021/day2/day2.hs
parseMove :: String -> (String, Int)
parseMove move =
    let [action, number] = words move
    in (action, read number :: Int)

part1 :: [(String, Int)] -> (Int, Int)
part1 [] = (0, 0)
part1 ((a, n):ms)
    | a == "forward" = (x + n, y)
    | a == "down" = (x, y + n)
    | a == "up" = (x, y - n)
    | otherwise = (x, y)
    where (x, y) = part1 ms

updateAim :: String -> Int -> Int -> Int
updateAim action aim n
    | action == "down" = aim + n
    | action == "up" = aim - n
    | otherwise = aim

part2 :: [(String, Int)] -> Int -> (Int, Int)
part2 [] _ = (0, 0)
part2 ((action, n):ms) a
    | action == "forward" = (x + n, y + a * n)
    | action == "down" = (x, y)
    | action == "up" = (x, y)
    | otherwise = (x, y)
    where 
        aim = updateAim action a n
        (x, y) = part2 ms aim

main :: IO ()
main = do
    input <- getContents
    let moves = map parseMove (lines input)
    let (x, y) = part1 moves
    putStrLn $ "Part 1: " ++ show (x * y)
    let (x, y) = part2 moves 0
    putStrLn $ "Part 2: " ++ show (x * y)

File: 2022/day2.py
with open('day2.txt', 'r') as f:
    data = f.read()

moves = {'X': 1, 'Y': 2, 'Z': 3}
translate = {'A': 'X', 'B': 'Y', 'C': 'Z'}

part1 = 0
part2 = 0
for line in data.split('\n'):
    if not line:    
        continue

    a, b = line.split()
    
    part1 += moves[b]
    part1 += {
        ('A', 'X'): 3,
        ('B', 'X'): 0,
        ('C', 'X'): 6,
        ('A', 'Y'): 6,
        ('B', 'Y'): 3,
        ('C', 'Y'): 0,
        ('A', 'Z'): 0,
        ('B', 'Z'): 6,
        ('C', 'Z'): 3
    }[(a, b)]

    if b == 'X':
        if a == 'A':
            move = 'Z'
        elif a == 'B':
            move = 'X'
        elif a == 'C':
            move = 'Y'
        part2 += 0
    elif b == 'Y':
        move = translate[a]
        part2 += 3
    else:
        if a == 'A':
            move = 'Y'
        elif a == 'B':
            move = 'Z'
        elif a == 'C':
            move = 'X'
        part2 += 6
    part2 += moves[move]
    
print('Part 1:', part1)
print('Part 2:', part2)
